### PR TITLE
Swift: Add path injection query

### PR DESCRIPTION
**Pull request by atorralba**
Adds a query for detecting path injection problems in Swift applications. See comments below for some caveats.

### CODE REVIEW

File: swift/ql/lib/codeql/swift/frameworks/StandardLibrary/NsUrl.qll [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+/**
+ * A model for `NSURL` members that permit taint flow.
+ */
+private class NsUrlSummaries extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        ";NSURL;true;init(string:);(String);;Argument[0];ReturnValue;taint",
+        // TODO
```
Comment (atorralba) [inline swift/ql/lib/codeql/swift/frameworks/StandardLibrary/NsUrl.qll:14]: `NSURL` should be properly modeled in its own PR.

File: swift/ql/lib/codeql/swift/security/PathInjection.qll [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+   */
+  abstract predicate step(DataFlow::Node node1, DataFlow::Node node2);
+}
+
+private class DefaultPathInjectionSink extends PathInjectionSink {
+  DefaultPathInjectionSink() { sinkNode(this, "path-injection") }
+}
+
+private class DefaultPathInjectionSanitizer extends PathInjectionSanitizer {
+  DefaultPathInjectionSanitizer() { none() } // TODO: Implement a proper path sanitizer
```
Comment (atorralba) [inline swift/ql/lib/codeql/swift/security/PathInjection.qll:32]: A proper path sanitizer can become quite complex (see [the Java implementation](https://github.com/github/codeql/blob/7456f3750d36f56d413839636513849c980b5c97/java/ql/lib/semmle/code/java/security/PathSanitizer.qll) for a reference).

I think it would make review easier if we created that in its own PR. But let me know if you disagree and want to see that in this PR (as of now, the recommendation made in the QHelp (`PathInjectionGood.swift` sample) would not actually make alerts disappear, which is always a bad user experience).

File: swift/ql/lib/codeql/swift/frameworks/StandardLibrary/NsUrl.qll (hunk 1 above)
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/frameworks/StandardLibrary/NsUrl.qll:14]: And URL for that matter.

File: swift/ql/lib/codeql/swift/security/PathInjection.qll (hunk 2 above)
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/security/PathInjection.qll:32]: What kind of level of false positives are we likely to see with no sanitizer here?  Right now lack of results is a much higher priority for Swift than false positives (within reason) and I've deferred other similar work until well into 2023 for this reason.

On the flip side its nice to get things finished while they're on your mind.  And I agree that giving advice in the `qhelp` that won't make the alert go away is a problem.

File: swift/ql/lib/codeql/swift/frameworks/StandardLibrary/NsUrl.qll [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+      ]
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/frameworks/StandardLibrary/NsUrl.qll:15]: I'm dismissing this because of the `\\ TODO`.

File: swift/ql/lib/codeql/swift/security/PathInjection.qll (hunk 2 above)
Comment (atorralba) [reply swift/ql/lib/codeql/swift/security/PathInjection.qll:32]: > What kind of level of false positives are we likely to see with no sanitizer here?

I'd need to run the query on a decent amount of projects to be sure, but intuitively, any app doing filesystem operations based on user input would get flagged, regardless of how well they validate the input. So it depends on how many apps do the risky thing (user input into filesystem sink), and how many of those sanitize/validate well.

> On the flip side its nice to get things finished while they're on your mind. And I agree that giving advice in the qhelp that won't make the alert go away is a problem.

As a compromise, we could write a very basic sanitizer for the moment that does the bare minimum to detect at least the simplest sanitization/validation logic (including the GOOD case in the `qhelp`), and defer the rest until we have a proper Guards library.

Comment (geoffw0) [reply swift/ql/lib/codeql/swift/security/PathInjection.qll:32]: Sounds like a plan. :+1:

File: swift/ql/lib/codeql/swift/frameworks/StandardLibrary/NsUrl.qll (hunk 1 above)
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/frameworks/StandardLibrary/NsUrl.qll:14]: Issue here: https://github.com/github/codeql-c-team/issues/1429

File: swift/ql/lib/codeql/swift/dataflow/ExternalFlow.qll [hunk 4]
```diff
@@ -82,10 +82,12 @@ private module Frameworks {
   private import codeql.swift.frameworks.StandardLibrary.Data
   private import codeql.swift.frameworks.StandardLibrary.InputStream
   private import codeql.swift.frameworks.StandardLibrary.NSData
+  private import codeql.swift.frameworks.StandardLibrary.NsUrl
   private import codeql.swift.frameworks.StandardLibrary.String
   private import codeql.swift.frameworks.StandardLibrary.Url
   private import codeql.swift.frameworks.StandardLibrary.UrlSession
   private import codeql.swift.frameworks.StandardLibrary.WebView
+  private import codeql.swift.security.PathInjection
```
Comment (geoffw0) [inline swift/ql/lib/codeql/swift/dataflow/ExternalFlow.qll:90]: It feels like this import belongs be somewhere else?

Comment (atorralba) [reply swift/ql/lib/codeql/swift/dataflow/ExternalFlow.qll:90]: This is needed for the MaD sinks present in `PathInjection.qll` to be added to the global scope, so that flow paths for all queries are only computed once. Otherwise, running the path injection query would require (AFAIK) some recomputation, impacting performance.

File: swift/ql/lib/codeql/swift/security/PathInjection.qll (hunk 2 above)
Comment (atorralba) [reply swift/ql/lib/codeql/swift/security/PathInjection.qll]: Done in 52ebf66. It took a while because I was banging my head against the wall until I figured out https://github.com/github/codeql/pull/11469 was needed.

File: swift/ql/test/query-tests/Security/CWE-022/testPathInjection.swift [hunk 5]
```diff
@@ -107,76 +113,88 @@ func test() {
...
+    let _ = fm.replaceItemAtURL(originalItemURL: safeNsUrl, withItemAtURL: remoteNsUrl, backupItemName: nil, options: []) // $ hasPathInjection=110
+}
+
+func testSanitizers() {
+    let remoteString = String(contentsOf: URL(string: "http://example.com/")!)
+
+    let fm = FileManager()
+
+    let filePath = FilePath(stringLiteral: remoteString)
+    if (filePath.lexicallyNormalized().starts(with: FilePath(stringLiteral: "/safe"))) {
```
Comment (MathiasVP) [inline swift/ql/test/query-tests/Security/CWE-022/testPathInjection.swift:196]: Note: We _do_ have https://github.com/github/codeql/blob/main/swift/ql/lib/codeql/swift/controlflow/BasicBlocks.qll#L272 which you could use to remove this FN. Although, I'm not sure how much effort we want to put into this since we're likely to rewrite this once we have a shared guards library.

Comment (atorralba) [reply swift/ql/test/query-tests/Security/CWE-022/testPathInjection.swift:196]: Just in case this becomes a temporary solution that sticks around more than expected, I followed your advice and used dominance to remove the FN in bf023b0. Thanks for pointing me to `BasicBlocks.qll`.

**Review by geoffw0 — COMMENTED** [review]
I've just done a full review, this looks good to me.  I've suggested some minor changes.  We also need a docs review and to review the DCA run when it's done.  Happy for these things to be done in parallel.

File: swift/ql/lib/codeql/swift/security/PathInjection.qll [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        ";FileManager;true;linkItem(at:to:);;;Argument[0..1];path-injection",
+        ";FileManager;true;linkItem(atPath:toPath:);;;Argument[0..1];path-injection",
+        ";FileManager;true;destinationOfSymbolicLink(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;fileExists(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;fileExists(atPath:isDirectory:);;;Argument[0];path-injection",
+        ";FileManager;true;isReadableFile(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;isWritableFile(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;isDeletableFile(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;componentsToDisplay(forPath:);;;Argument[0];path-injection",
+        ";FileManager;true;displayName(atPath:);;;Argument[0];path-injection",
```
[the remark this answers is not in the document]
Comment (atorralba) [reply swift/ql/lib/codeql/swift/security/PathInjection.qll:98]: Yes, I was doubtful about adding certain sinks. My logic ended up being that if the output of these methods was used in a security-relevant check, an attacker could still cause some compromise by altering which file gets used. But in retrospect you're probably right: that is probably very FP-prone, and we should remove some.

I'd remove the following rows, let me know what you think:

```ql
";FileManager;true;isReadableFile(atPath:);;;Argument[0];path-injection",
";FileManager;true;isWritableFile(atPath:);;;Argument[0];path-injection",
";FileManager;true;isDeletableFile(atPath:);;;Argument[0];path-injection",
";FileManager;true;attributesOfItem(atPath:);;;Argument[0];path-injection",
";FileManager;true;attributesOfFileSystem(forPath:);;;Argument[0];path-injection",
";FileManager;true;setAttributes(_:ofItemAtPath:);;;Argument[1];path-injection", // arguably
";FileManager;true;getRelationship(_:ofDirectoryAt:toItemAt:);;;Argument[1..2];path-injection",
";FileManager;true;getRelationship(_:of:in:toItemAt:);;;Argument[3];path-injection",
";FileManager;true;NSHFSTypeOfFile(_:);;;Argument[0];path-injection",

File: swift/ql/lib/codeql/swift/security/PathInjection.qll [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        ";FileManager;true;subpaths(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;createDirectory(at:withIntermediateDirectories:attributes:);;;Argument[0];path-injection",
+        ";FileManager;true;createDirectory(atPath:withIntermediateDirectories:attributes:);;;Argument[0];path-injection",
+        ";FileManager;true;createFile(atPath:contents:attributes:);;;Argument[0];path-injection",
+        ";FileManager;true;removeItem(at:);;;Argument[0];path-injection",
+        ";FileManager;true;removeItem(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;trashItem(at:resultingItemURL:);;;Argument[0];path-injection",
+        ";FileManager;true;replaceItem(at:withItemAt:backupItemName:options:resultingItemURL:);;;Argument[0..1];path-injection",
+        ";FileManager;true;replaceItemAt(_:withItemAt:backupItemName:options:);;;Argument[0..1];path-injection",
+        ";FileManager;true;replaceItemAt(_:withItemAt:backupItemName:options:resultingItemURL:);;;Argument[0..1];path-injection",
```
[the remark this answers is not in the document]
Comment (atorralba) [reply swift/ql/lib/codeql/swift/security/PathInjection.qll:82]: Good catch, that was a copy-paste error.

File: swift/ql/lib/codeql/swift/security/PathInjection.qll [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        ";FileManager;true;createSymbolicLink(at:withDestinationURL:);;;Argument[0..1];path-injection",
+        ";FileManager;true;createSymbolicLink(atPath:withDestinationPath:);;;Argument[0..1];path-injection",
+        ";FileManager;true;linkItem(at:to:);;;Argument[0..1];path-injection",
+        ";FileManager;true;linkItem(atPath:toPath:);;;Argument[0..1];path-injection",
+        ";FileManager;true;destinationOfSymbolicLink(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;fileExists(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;fileExists(atPath:isDirectory:);;;Argument[0];path-injection",
+        ";FileManager;true;isReadableFile(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;isWritableFile(atPath:);;;Argument[0];path-injection",
+        ";FileManager;true;isDeletableFile(atPath:);;;Argument[0];path-injection",
```
[the remark this answers is not in the document]
Comment (atorralba) [reply swift/ql/lib/codeql/swift/security/PathInjection.qll:96]: See my comment below, I think this set of sinks should be removed anyway.

File: swift/ql/lib/codeql/swift/security/PathInjection.qll (hunk 6 above)
[the remark this answers is not in the document]
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/security/PathInjection.qll:97]: `isReadableFile`, `isWritableFile` and the others I'm not so worried about FPs from.  Usually you check a file is readable before you read it, writable before you write it (etc), and I wouldn't want remote data to be controlling that.  I would probably keep them in the results personally, but I'm not too worried either way.

I would keep `setAttributes` - seems like it could be dangerous.

The case of `displayName` and `componentsToDisplay` is special - these are likely to be used to display something to a user, e.g. status or a question, maybe even an error.  It seems to me that we shouldn't flag merely displaying a remote-controlled path unless something is being done with it.

[the remark this answers is not in the document]
Comment (atorralba) [reply swift/ql/lib/codeql/swift/security/PathInjection.qll:97]: >  Usually you check a file is readable before you read it, writable before you write it (etc),

But then the appropriate sinks for read/write would be used anyway, and I'd prefer not to duplicate alerts around the same functionality.

> I would keep setAttributes - seems like it could be dangerous.

👍

[the remark this answers is not in the document]
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/security/PathInjection.qll:97]: > But then the appropriate sinks for read/write would be used anyway, and I'd prefer not to duplicate alerts around the same functionality.

Good point.

**Comment by geoffw0** [comment]
DCA looks good to me.  I wouldn't be too concerned about the analysis time regressions as the overall time did not regress (I think there's just quite a lot of wobble, on certain projects at least).  The "tuple sums, per source" results are weird, but not really actionable - and notably "tuple sums, per source and predicate" does not flag anything.  Stage timings also doesn't flag any regressions.  I think there's nothing to worry about here.

File: swift/ql/lib/codeql/swift/security/PathInjection.qll (hunk 6 above)
[the remark this answers is not in the document]
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/security/PathInjection.qll]: I'm happy with the changes you've made.

**Comment by geoffw0** [comment]
I think we're just waiting on a doc review now. :+1:

**Comment by MathiasVP** [comment]
> DCA looks good to me. I wouldn't be too concerned about the analysis time regressions as the overall time did not regress (I think there's just quite a lot of wobble, on certain projects at least). The "tuple sums, per source" results are weird, but not really actionable - and notably "tuple sums, per source and predicate" does not flag anything. Stage timings also doesn't flag any regressions. I think there's nothing to worry about here.

I think "tuple sums, per source" has increased because we're adding a _whole new query_. So the reason there are more tuples is that we're evaluating a lot more predicates now.

**Review by subatoi — COMMENTED** [review]
Hi, I've left a few comments from Docs 👋 

Nothing major but just trying to align it with our existing style as much as possible, and give it a bit of a trim. Please let me know if I can expand on anything and I'll be happy to take another look after you've read them

File: swift/ql/src/queries/Security/CWE-022/PathInjection.qhelp [hunk 9]
```diff
@@ -0,0 +1,58 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+
+<overview>
+<p>Accessing paths controlled by users can allow an attacker to access unexpected resources. This 
+can result in sensitive information being revealed or deleted, or an attacker being able to influence
+behavior by modifying unexpected files.</p>
```
[the remark this answers is not in the document]
Comment (atorralba) [reply swift/ql/src/queries/Security/CWE-022/PathInjection.qhelp:9]: I guess the "or an attacker being able to influence behavior by modifying unexpected files" part is lost with your change, but it isn't a big deal. No strong opinion about this, so if you think it's better to keep it short, let's do it 😄.

File: swift/ql/src/queries/Security/CWE-022/PathInjection.qhelp [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+<li>Do not rely on simply replacing problematic sequences such as "../". For example, after applying this filter to
+".../...//" the resulting string would still be "../".</li>
+<li>Ideally use a whitelist of known good patterns.</li>
+</ul>
+
+</recommendation>
+
+<example>
+<p>
+In the first example, a file name is read from an HTTP request and then used to access a file.
```
[the remark this answers is not in the document]
Comment (atorralba) [reply swift/ql/src/queries/Security/CWE-022/PathInjection.qhelp:29]: Sorry, could you clarify? Is it that the first two examples being in separate paragraphs but using the same `sample` looks weird? Or is it that we should change the "first", "second", and "third" words for "the following"?

Right now the structure is:

In the first (bad) example, ...
In the second (bad) example, ...

```
<Bad example 1>
<Bad example 2>
```

In the third (good) example, ...

```
<Good example>
```

**Comment by atorralba** [comment]
Most suggestions applied @subatoi, thanks for the review! One question remains.

I think it's worth mentioning that, as with previous queries, the `qhelp` was copy-pasted from the same query in other languages, so we should probably apply the same changes in all of them after this.

File: swift/ql/src/queries/Security/CWE-022/PathInjection.qhelp (hunk 10 above)
[the remark this answers is not in the document]
Comment (subatoi) [reply swift/ql/src/queries/Security/CWE-022/PathInjection.qhelp:29]: It's minor feedback on my part and sorry for the confusion but yes, I'd say this was true to some extent:

> Is it that the first two examples being in separate paragraphs but using the same sample looks weird

I found this pattern conceptually clearer:

https://github.com/github/codeql/pull/11238/files#diff-b7b86a6525317add19bcf4c899b4c5a6c645e92267e250c3195cebfac9f2d5b1R33

**Comment by subatoi** [comment]
Thanks @atorralba, I'll tag @felicitymay on that as I think we'll standardise it later when it comes over to docs.github.com but I'm not 100% sure of the timelines

File: swift/ql/src/queries/Security/CWE-022/PathInjection.qhelp (hunk 10 above)
[the remark this answers is not in the document]
Comment (subatoi) [reply swift/ql/src/queries/Security/CWE-022/PathInjection.qhelp:29]: Let me resolve this thread and add a suggestion, and we can go from there. Sorry for the confusion 🙏

File: swift/ql/src/queries/Security/CWE-022/PathInjection.qhelp [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+<sample src="PathInjectionBad.swift" />
+
+<p>
+In the third example, the path used to access the file system is normalized <em>before</em> being checked against a
+known prefix. This ensures that regardless of the user input, the resulting path is safe.
+</p>
+
+<sample src="PathInjectionGood.swift" />
+
```
Comment (subatoi) [inline swift/ql/src/queries/Security/CWE-022/PathInjection.qhelp:50]:
```suggestion
<p>
The following code shows two bad examples. 

In the first, a file name is read from an HTTP request and then used to access a file. In this case, a malicious response could include a file name that is an absolute path, such as
<code>"/Applications/(current_application)/Documents/sensitive.data"</code>.
</p>

<p>
In the second bad example, it appears that the user is restricted to opening a file within the
<code>"/Library/Caches"</code> home directory. In this case, a malicious response could contain a file name containing
special characters. For example, the string <code>"../../Documents/sensitive.data"</code> will result in the code
reading the file located at <code>"/Applications/(current_application)/Library/Caches/../../Documents/sensitive.data"</code>, 
which contains users' sensitive data. This file may then be made accessible to an attacker, giving them access to all this data.
</p>

<sample src="PathInjectionBad.swift" />

<p>
In the following (good) example, the path used to access the file system is normalized <em>before</em> being checked against a
known prefix. This ensures that regardless of the user input, the resulting path is safe.
</p>

<sample src="PathInjectionGood.swift" />
```suggestion

Comment (subatoi) [reply swift/ql/src/queries/Security/CWE-022/PathInjection.qhelp:50]: @atorralba Let me know what you think and sorry again for the confusion. Only really important thing here is the tiny typo tbh (`accesible`).

**Comment by felicitymay** [comment]
Thanks for the updates and the ping.

> I think we'll standardise it later when it comes over to docs.github.com but I'm not 100% sure of the timelines

It would definitely be good to standardise the language in similar help files. We don't currently have an issue open to do this work, and I think that it will be a while before we can find time.

File: swift/ql/src/queries/Security/CWE-022/PathInjection.qhelp (hunk 11 above)
Comment (atorralba) [reply swift/ql/src/queries/Security/CWE-022/PathInjection.qhelp]: Thanks for the suggestion @subatoi! Change applied 😄

**Review by geoffw0 — DISMISSED** [review]

File: swift/ql/src/queries/Security/CWE-022/PathInjection.qhelp [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+<example>
+<p>
+The following code shows two bad examples. 
+</p>
+
+<sample src="PathInjectionBad.swift" />
+
+<p>
+In the first, a file name is read from an HTTP request and then used to access a file. In this case, a malicious response could include a file name that is an absolute path, such as
```
Comment (subatoi) [inline swift/ql/src/queries/Security/CWE-022/PathInjection.qhelp:35]:
```suggestion
In the first bad example, a file name is read from an HTTP request and then used to access a file. In this case, a malicious response could include a file name that is an absolute path, such as
```

Not urgent, just makes it a tiny bit clearer. Apologies for not suggesting originally.

**Review by subatoi — DISMISSED** [review]
Thank you! Two non-urgent comments that are only really worth applying if you don't need to rebuild everything again / get reapprovals.

**Review by subatoi — APPROVED** [review]

**Review by geoffw0 — APPROVED** [review]
